test/distributed/checkpoint/test_file_system_checkpoint.py

```diff
@@ -0,0 +1,496 @@
+                if s0 == s1:
+                    continue
+
+                if dist.get_rank() == 0:
+                    shutil.rmtree(path, ignore_errors=True)
+                    os.makedirs(path)
+                dist.barrier()
+
+                model_to_save = MyShardedModel3(s0)
+                model_to_save._register_state_dict_hook(state_dict_hook)
+                state_dict_to_save = model_to_save.state_dict()
+
+                fs_writer = FileSystemWriter(path=path)
+                save_state_dict(state_dict=state_dict_to_save, storage_writer=fs_writer)
+
+                dist.barrier()
+
+                model_to_load = MyShardedModel3(s1)
+                model_to_load._register_state_dict_hook(state_dict_hook)
+                state_dict_to_load_to = model_to_load.state_dict()
+                dist.barrier()
+
+                fs_reader = FileSystemReader(path=path)
+                load_state_dict(
+                    state_dict=state_dict_to_load_to, storage_reader=fs_reader
+                )
+
+                dist.barrier()
+                store_tensor = self.load_tensor(model_to_save.sharded_tensor)
+                dist.barrier()
+                load_tensor = self.load_tensor(model_to_load.sharded_tensor)
+
+                if dist.get_rank() == 0:
+                    self.assertTrue(
+                        torch.allclose(store_tensor, load_tensor), msg=f"{s0} vs {s1}"
+                    )
+
+    @with_comms(init_rpc=False)
+    @skip_if_lt_x_gpu(2)
+    @requires_nccl()
+    def test_load_rowwise_to_colwise(self) -> None:
+        print(f"{dist.get_rank()}---1")
+        path = self.get_file_path()
+        self.assertEqual(self.world_size, dist.get_world_size())
+        print(f"{dist.get_rank()}---2")
+
+
+        # pyre-fixme [28]: Unexpected keyword argument `dim` to call `dist._sharding_spec.api.ChunkShardingSpec.__init__`.
+        src_spec = ChunkShardingSpec(
+            dim=0,
+            placements=[
+                "rank:0/cuda:0",
+                "rank:1/cuda:1",
+            ],
+        )
+
+        # pyre-fixme [28]: Unexpected keyword argument `dim` to call `dist._sharding_spec.api.ChunkShardingSpec.__init__`.
+        dst_spec = ChunkShardingSpec(
+            dim=1,
+            placements=[
+                "rank:0/cuda:0",
+                "rank:1/cuda:1",
+            ],
+        )
+
+        if dist.get_rank() == 0:
+            shutil.rmtree(path, ignore_errors=True)
+            os.makedirs(path)
+
+        model_to_save = MyShardedModel3(src_spec)
+        model_to_save._register_state_dict_hook(state_dict_hook)
+        state_dict_to_save = model_to_save.state_dict()
+
+        fs_writer = FileSystemWriter(path=path)
+        save_state_dict(state_dict=state_dict_to_save, storage_writer=fs_writer)
+
+        model_to_load = MyShardedModel3(dst_spec)
+        model_to_load._register_state_dict_hook(state_dict_hook)
+        state_dict_to_load_to = model_to_load.state_dict()
+
```

Comment:
don't we need a barrier here? what if some rank (i.e. rank 0) finishs saving the state_dict and some not, then rank 0 starts loading the state_dict but rank 1 still not yet finishs saving the local shards to the directory? This might still work with no-resharding, but with resharding, I think it might cause some problem?

Reply:
I'm not sure what the barrier would be coordinating here.

Load-time resharding is a pure-local operation where each rank independently plans how to load from the checkpoint.

Reply:
Filed a follow up issue on ensuring load/save has reasonable SPMD coordination behavior

Reply:
https://github.com/pytorch/pytorch/issues/76395